### DISCUSSION TITLE
Undefined symbols for architecture x86_64

### DISCUSSION THREAD

**Discussion by xchengood** [neperfepx/neper#424] [General]
Hi Neper experts or users,

I am new to Neper. One problem below occurs in installation on the macOS system. Could anyone help me solve this issue? Thank you.

> [100%] Built target neper_m
[100%] Building C object CMakeFiles/neper.dir/neper.c.o
[100%] Linking CXX executable neper
Undefined symbols for architecture x86_64:
  "___kmpc_critical", referenced from:
      _.omp_outlined. in libneper_m.a(nem_meshing_2d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_3d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_str2.c.o)
      _nem_meshing_2d_face in libneper_m.a(nem_meshing_2d2.c.o)
      _.omp_outlined. in libneper_m.a(nem_transport3.c.o)
      _nem_meshing_2d_face_mesh in libneper_m.a(nem_meshing_2d3.c.o)
      _.omp_outlined. in libneper_t.a(net_transform_tesr_tesrtranslate_bounds1.c.o)
      ...
  "___kmpc_dispatch_init_4", referenced from:
      _.omp_outlined. in libneper_t.a(net_utils1.c.o)
      _.omp_outlined..70 in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined..71 in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_2d1.c.o)
      _.omp_outlined..3 in libneper_m.a(nem_meshing_2d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_3d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_str2.c.o)
      ...
  "___kmpc_dispatch_init_4u", referenced from:
      _.omp_outlined..1 in libneper_t.a(net_ori_uniform4.cpp.o)
      _.omp_outlined. in libneper_t.a(net_tess_opt_comp_objective_fval_gen_stat2.c.o)
      _.omp_outlined..1 in libneper_t.a(net_tess_opt_comp_objective_fval_gen_stat2.c.o)
      _.omp_outlined..2 in libneper_t.a(net_tess_opt_comp_objective_fval_gen_stat2.c.o)
  "___kmpc_dispatch_next_4", referenced from:
      _.omp_outlined. in libneper_t.a(net_utils1.c.o)
      _.omp_outlined..70 in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined..71 in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_2d1.c.o)
      _.omp_outlined..3 in libneper_m.a(nem_meshing_2d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_3d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_str2.c.o)
      ...
  "___kmpc_dispatch_next_4u", referenced from:
      _.omp_outlined..1 in libneper_t.a(net_ori_uniform4.cpp.o)
      _.omp_outlined. in libneper_t.a(net_tess_opt_comp_objective_fval_gen_stat2.c.o)
      _.omp_outlined..1 in libneper_t.a(net_tess_opt_comp_objective_fval_gen_stat2.c.o)
      _.omp_outlined..2 in libneper_t.a(net_tess_opt_comp_objective_fval_gen_stat2.c.o)
  "___kmpc_end_critical", referenced from:
      _.omp_outlined. in libneper_m.a(nem_meshing_2d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_3d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_str2.c.o)
      _nem_meshing_2d_face in libneper_m.a(nem_meshing_2d2.c.o)
      _.omp_outlined. in libneper_m.a(nem_transport3.c.o)
      _nem_meshing_2d_face_mesh in libneper_m.a(nem_meshing_2d3.c.o)
      _.omp_outlined. in libneper_t.a(net_transform_tesr_tesrtranslate_bounds1.c.o)
      ...
  "___kmpc_for_static_fini", referenced from:
      _.omp_outlined. in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined..69 in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined. in libneut.a(neut_ori1.c.o)
      _.omp_outlined. in libneut.a(neut_tesr_op1.c.o)
      _.omp_outlined..27 in libneut.a(neut_tesr_op1.c.o)
      _.omp_outlined..28 in libneut.a(neut_tesr_op1.c.o)
      _.omp_outlined. in libneut.a(neut_data_gen1.cpp.o)
      ...
  "___kmpc_for_static_init_4", referenced from:
      _.omp_outlined. in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined..69 in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined. in libneut.a(neut_ori1.c.o)
      _.omp_outlined. in libneut.a(neut_tesr_op1.c.o)
      _.omp_outlined..27 in libneut.a(neut_tesr_op1.c.o)
      _.omp_outlined..28 in libneut.a(neut_tesr_op1.c.o)
      _.omp_outlined. in libneut.a(neut_data_gen1.cpp.o)
      ...
  "___kmpc_for_static_init_4u", referenced from:
      _.omp_outlined. in libneper_t.a(net_ori_uniform3.cpp.o)
      _.omp_outlined..1 in libneper_t.a(net_ori_uniform3.cpp.o)
      _.omp_outlined..2 in libneper_t.a(net_ori_uniform3.cpp.o)
      _.omp_outlined. in libneper_t.a(net_ori_uniform4.cpp.o)
      _.omp_outlined..2 in libneper_t.a(net_ori_uniform4.cpp.o)
      _.omp_outlined..3 in libneper_t.a(net_ori_uniform4.cpp.o)
  "___kmpc_fork_call", referenced from:
      _neper_head in neper.c.o
      _net_tess_tesr in libneper_t.a(net_utils1.c.o)
      _neut_mesh_entity_expr_val in libneut.a(neut_mesh_gen.c.o)
      _neut_mesh_eltdata_elsetdata in libneut.a(neut_mesh_gen.c.o)
      _neut_mesh_eltdata_elsetdata_ori in libneut.a(neut_mesh_gen.c.o)
      _neut_mesh_eltdata_elsetdata_oridis in libneut.a(neut_mesh_gen.c.o)
      _neut_ori_f_norm in libneut.a(neut_ori1.c.o)
      ...
  "___kmpc_global_thread_num", referenced from:
      _nem_meshing_2d_face in libneper_m.a(nem_meshing_2d2.c.o)
      _nem_meshing_2d_face_mesh in libneper_m.a(nem_meshing_2d3.c.o)
      net_polycomp_cells_updatecell(POLY, SEEDSET, nanoflann::KDTreeSingleIndexDynamicAdaptor<nanoflann::L2_Simple_Adaptor<double, NFCLOUD, double>, NFCLOUD, 3, unsigned long>**, int*, int, POLY**, TDYN*) in libneper_t.a(net_polycomp3.cpp.o)
      _nem_meshing_2d_face_mesh_gmsh in libneper_m.a(nem_meshing_2d_face_mesh_gmsh1.c.o)
      net_polycomp_poly(POLY, SEEDSET, nanoflann::KDTreeSingleIndexDynamicAdaptor<nanoflann::L2_Simple_Adaptor<double, NFCLOUD, double>, NFCLOUD, 3, unsigned long>**, int*, int, POLY*, TDYN*) in libneper_t.a(net_polycomp_poly1.cpp.o)
  "_omp_get_num_threads", referenced from:
      _.omp_outlined. in neper.c.o
  "_omp_get_thread_num", referenced from:
      _.omp_outlined. in neper.c.o
      _nem_mesh_3d_gmsh in libneper_m.a(nem_mesh_gmsh1.c.o)
      _nem_meshing_2d_face_mesh_gmsh in libneper_m.a(nem_meshing_2d_face_mesh_gmsh1.c.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [neper] Error 1
make[1]: *** [CMakeFiles/neper.dir/all] Error 2
make: *** [all] Error 2

**Comment by mkasemer** [comment]
What compiler are you using, Xcode? Have you tried turning OMP off (using `ccmake`, for example)?

Note that the default compilers on MacOS do not have native OMP support. You will have to utilize other compilers, such as LLVM (via Brew), to build with OMP support.

**Reply by xchengood** [reply]
Hi @mkasemer. Thank you for your advice. Yes, I am using Xcode. I have no idea how to turn omp off in the right way. I just use **ccmake -DHAVE_OpenMP={OFF}**.

I already install LLVM and libomp via brew.

Also, I notice I am using one called AppleClang and do not know what exactly is.

**Comment by rquey** [comment]
@mkasemer knows about compiling on macOS much more than me, but it should start by `cmake  -DHAVE_OpenMP=OFF ..`, and then `make`. https://neper.info/doc/introduction.html#installing-neper

**Reply by xchengood** [reply]
Hi @rquey, follow your advice. I still got the same errors shown below.

> [100%] Building C object neper_m/CMakeFiles/neper_m.dir/neper_m.c.o
[100%] Linking CXX static library libneper_m.a
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libneper_m.a(nem_elsets2.c.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libneper_m.a(nem_elsets2.c.o) has no symbols
[100%] Built target neper_m
[100%] Building C object CMakeFiles/neper.dir/neper.c.o
[100%] Linking CXX executable neper
Undefined symbols for architecture x86_64:
  "___kmpc_critical", referenced from:
      _.omp_outlined. in libneper_m.a(nem_meshing_2d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_3d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_str2.c.o)
      _nem_meshing_2d_face in libneper_m.a(nem_meshing_2d2.c.o)
      _.omp_outlined. in libneper_m.a(nem_transport3.c.o)
      _nem_meshing_2d_face_mesh in libneper_m.a(nem_meshing_2d3.c.o)
      _.omp_outlined. in libneper_t.a(net_transform_tesr_tesrtranslate_bounds1.c.o)
      ...
  "___kmpc_dispatch_init_4", referenced from:
      _.omp_outlined. in libneper_t.a(net_utils1.c.o)
      _.omp_outlined..70 in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined..71 in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_2d1.c.o)
      _.omp_outlined..3 in libneper_m.a(nem_meshing_2d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_3d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_str2.c.o)
      ...
  "___kmpc_dispatch_init_4u", referenced from:
      _.omp_outlined..1 in libneper_t.a(net_ori_uniform4.cpp.o)
      _.omp_outlined. in libneper_t.a(net_tess_opt_comp_objective_fval_gen_stat2.c.o)
      _.omp_outlined..1 in libneper_t.a(net_tess_opt_comp_objective_fval_gen_stat2.c.o)
      _.omp_outlined..2 in libneper_t.a(net_tess_opt_comp_objective_fval_gen_stat2.c.o)
  "___kmpc_dispatch_next_4", referenced from:
      _.omp_outlined. in libneper_t.a(net_utils1.c.o)
      _.omp_outlined..70 in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined..71 in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_2d1.c.o)
      _.omp_outlined..3 in libneper_m.a(nem_meshing_2d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_3d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_str2.c.o)
      ...
  "___kmpc_dispatch_next_4u", referenced from:
      _.omp_outlined..1 in libneper_t.a(net_ori_uniform4.cpp.o)
      _.omp_outlined. in libneper_t.a(net_tess_opt_comp_objective_fval_gen_stat2.c.o)
      _.omp_outlined..1 in libneper_t.a(net_tess_opt_comp_objective_fval_gen_stat2.c.o)
      _.omp_outlined..2 in libneper_t.a(net_tess_opt_comp_objective_fval_gen_stat2.c.o)
  "___kmpc_end_critical", referenced from:
      _.omp_outlined. in libneper_m.a(nem_meshing_2d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_3d1.c.o)
      _.omp_outlined. in libneper_m.a(nem_meshing_str2.c.o)
      _nem_meshing_2d_face in libneper_m.a(nem_meshing_2d2.c.o)
      _.omp_outlined. in libneper_m.a(nem_transport3.c.o)
      _nem_meshing_2d_face_mesh in libneper_m.a(nem_meshing_2d3.c.o)
      _.omp_outlined. in libneper_t.a(net_transform_tesr_tesrtranslate_bounds1.c.o)
      ...
  "___kmpc_for_static_fini", referenced from:
      _.omp_outlined. in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined..69 in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined. in libneut.a(neut_ori1.c.o)
      _.omp_outlined. in libneut.a(neut_tesr_op1.c.o)
      _.omp_outlined..27 in libneut.a(neut_tesr_op1.c.o)
      _.omp_outlined..28 in libneut.a(neut_tesr_op1.c.o)
      _.omp_outlined. in libneut.a(neut_data_gen1.cpp.o)
      ...
  "___kmpc_for_static_init_4", referenced from:
      _.omp_outlined. in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined..69 in libneut.a(neut_mesh_gen.c.o)
      _.omp_outlined. in libneut.a(neut_ori1.c.o)
      _.omp_outlined. in libneut.a(neut_tesr_op1.c.o)
      _.omp_outlined..27 in libneut.a(neut_tesr_op1.c.o)
      _.omp_outlined..28 in libneut.a(neut_tesr_op1.c.o)
      _.omp_outlined. in libneut.a(neut_data_gen1.cpp.o)
      ...
  "___kmpc_for_static_init_4u", referenced from:
      _.omp_outlined. in libneper_t.a(net_ori_uniform3.cpp.o)
      _.omp_outlined..1 in libneper_t.a(net_ori_uniform3.cpp.o)
      _.omp_outlined..2 in libneper_t.a(net_ori_uniform3.cpp.o)
      _.omp_outlined. in libneper_t.a(net_ori_uniform4.cpp.o)
      _.omp_outlined..2 in libneper_t.a(net_ori_uniform4.cpp.o)
      _.omp_outlined..3 in libneper_t.a(net_ori_uniform4.cpp.o)
  "___kmpc_fork_call", referenced from:
      _neper_head in neper.c.o
      _net_tess_tesr in libneper_t.a(net_utils1.c.o)
      _neut_mesh_entity_expr_val in libneut.a(neut_mesh_gen.c.o)
      _neut_mesh_eltdata_elsetdata in libneut.a(neut_mesh_gen.c.o)
      _neut_mesh_eltdata_elsetdata_ori in libneut.a(neut_mesh_gen.c.o)
      _neut_mesh_eltdata_elsetdata_oridis in libneut.a(neut_mesh_gen.c.o)
      _neut_ori_f_norm in libneut.a(neut_ori1.c.o)
      ...
  "___kmpc_global_thread_num", referenced from:
      _nem_meshing_2d_face in libneper_m.a(nem_meshing_2d2.c.o)
      _nem_meshing_2d_face_mesh in libneper_m.a(nem_meshing_2d3.c.o)
      net_polycomp_cells_updatecell(POLY, SEEDSET, nanoflann::KDTreeSingleIndexDynamicAdaptor<nanoflann::L2_Simple_Adaptor<double, NFCLOUD, double>, NFCLOUD, 3, unsigned long>**, int*, int, POLY**, TDYN*) in libneper_t.a(net_polycomp3.cpp.o)
      _nem_meshing_2d_face_mesh_gmsh in libneper_m.a(nem_meshing_2d_face_mesh_gmsh1.c.o)
      net_polycomp_poly(POLY, SEEDSET, nanoflann::KDTreeSingleIndexDynamicAdaptor<nanoflann::L2_Simple_Adaptor<double, NFCLOUD, double>, NFCLOUD, 3, unsigned long>**, int*, int, POLY*, TDYN*) in libneper_t.a(net_polycomp_poly1.cpp.o)
  "_omp_get_num_threads", referenced from:
      _.omp_outlined. in neper.c.o
  "_omp_get_thread_num", referenced from:
      _.omp_outlined. in neper.c.o
      _nem_mesh_3d_gmsh in libneper_m.a(nem_mesh_gmsh1.c.o)
      _nem_meshing_2d_face_mesh_gmsh in libneper_m.a(nem_meshing_2d_face_mesh_gmsh1.c.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [neper] Error 1
make[1]: *** [CMakeFiles/neper.dir/all] Error 2
make: *** [all] Error 2

**Reply by mkasemer** [reply]
So, I am unable to reproduce this error, and that could be due to so many things (hard to narrow down). I assumed it was due to OMP owing to the error messages, but @rquey would have to confirm.

I would also check to make sure that your path is correct using the different compiler, for instance, here is what I have set using LLVM  v12.0.0_1 installed via Brew:
```
export CC=/usr/local/opt/llvm/bin/clang
export CXX=/usr/local/opt/llvm/bin/clang++
export LDFLAGS="-L/usr/local/opt/llvm/lib"
export CPPFLAGS="-I/usr/local/opt/llvm/include"
```

**Reply by xchengood** [reply]
@mkasemer. Thank you. Let me follow your advice about LLVM to look at how is going.

**Reply by rquey** [reply]
Yes, it seems to complain about OMP, and also in nanoflann, which is one of the rare C11 parts, so, this is maybe also part of the problem...
